integration-test/src/test/groovy/de/gesellix/docker/client/DockerStackComposeIntegrationSpec.groovy

```diff
@@ -0,0 +1,162 @@
+    command: sh
+    deploy:
+      endpoint_mode: vip
+      mode: replicated
+      replicas: 1
+      restart_policy:
+        condition: on-failure
+        delay: 5s
+        max_attempts: 3
+        window: 120s
+    environment:
+      SOME_VAR: ${SOME_VAR}
+    networks:
+      - "my-subnet"
+    volumes:
+      - type: volume
+        source: shm
+        target: /dev/shm
+
+networks:
+  my-subnet:
+
+volumes:
+  shm:
+    driver_opts:
+      type: tmpfs
+      device: tmpfs
+      o: size=128m
+'''
+
+    def "deploy a new stack with compose file"() {
+        given:
+        dockerClient.initSwarm(newSwarmConfig())
+
+        def namespace = "new-stack"
+        def composeStream = composeFile.newInputStream()
+        def workingDir = composeFile.parent
+        def environment = [IMAGE_VERSION: '3.4', SOME_VAR: 'some-value']
+        def options = new DeployStackOptions()
+        def config = new DeployConfigReader(dockerClient).loadCompose(namespace, composeStream, workingDir, environment)
+
+        when:
+        dockerClient.stackDeploy(namespace, config, options)
+
+        then:
+        Thread.sleep(1000)
+        def tasks = dockerClient.stackPs(namespace).content
+        tasks.size() == config.services.size()
+
+        def spec = dockerClient.inspectService("${namespace}_service").content.Spec
+        def containerSpec = spec.TaskTemplate.ContainerSpec
+        spec.EndpointSpec == [Mode: 'vip']
+        spec.Labels == ['com.docker.stack.namespace': 'new-stack']
+        spec.Mode == [Replicated: [Replicas: 1]]
+        spec.Name == "${namespace}_service"
+        spec.Networks.Aliases == [['service']]
+        containerSpec.Image =~ "alpine:3.4(@sha256:[a-f0-9]{64})?"
+        containerSpec.Labels == ['com.docker.stack.namespace': 'new-stack']
+        containerSpec.Args == ['sh']
+        containerSpec.Env == ['SOME_VAR=some-value']
+        containerSpec.Mounts == [
+                [
+                        Type: 'volume',
+                        Source: 'new-stack_shm',
+                        Target: '/dev/shm',
+                        VolumeOptions: [Labels: ['com.docker.stack.namespace': 'new-stack']]
+                ]
+        ]
+        spec.TaskTemplate.RestartPolicy == [
+                Condition: 'on-failure',
+                Delay: 5000000000,
+                MaxAttempts: 3,
+                Window: 120000000000
+        ]
+
+        def networkInfo = waitAndRetrySilently { dockerClient.inspectNetwork("${namespace}_my-subnet") }
+        networkInfo.status.code == 200
+
+        def volumeInfo = waitAndRetrySilently { dockerClient.inspectVolume("${namespace}_shm") }
+        volumeInfo.status.code == 200
+
+        cleanup:
+        composeStream.close()
+        performSilently { dockerClient.stackRm(namespace) }
+        waitAndRetrySilently { dockerClient.rmVolume("${namespace}_shm") }
+        performSilently { dockerClient.leaveSwarm([force: true]) }
+    }
+
+    def setupSpec() {
```

Comment:
Please have a look at the other specs, I prefer an order like this:

- global fields for every test case
- setup method(s)
- cleanup method(s)
- test cases
- util methods/fixture factories

Reply:
I've moved the inlined YAML to a resource file, reordered fields/methods according to the above, and cleaned-up a bit. A future refactoring could be to move utility methods for delaying/retrying docker command-closures out to a separate utilities class - if those are useful also in other integration tests.